src/mem/ruby/structures/ALUFreeListArray.cc

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2023 The University of Wisconsin
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "mem/ruby/structures/ALUFreeListArray.hh"
+
+#include "base/intmath.hh"
+#include "mem/ruby/system/RubySystem.hh"
+#include "sim/cur_tick.hh"
+
+namespace gem5
+{
+
+namespace ruby
+{
+
+/*
+* Behavior:
+*   Requests will go through unless one/both of the following are met:
+*       - There have been more than [numALUs] requests in the current cycle
+*       - The same line has been accessed in the past accessLatency ticks
+*/
+
+ALUFreeListArray::ALUFreeListArray(unsigned int num_ALUs, Tick access_latency)
+{
+    this->numALUs = num_ALUs;
+    this->accessLatency = access_latency;
+}
+
+bool ALUFreeListArray::tryAccess(Addr addr)
+{
+    uint32_t accesses_this_tick = 0;
+
+    // Remove requests from the tail of the queue that occured more than
+    // accessLatency ticks ago
+    Tick oldestValidRecordStart = curTick() - this->accessLatency;
+
+    while (accessQueue.size() > 0 &&
+         (accessQueue.back().startTick < oldestValidRecordStart)) {
+        accessQueue.pop_back();
+    }
+
+    for (AccessRecord& record : accessQueue) {
+        // Block access if we would be using more ALUs than we have in a
+        // single tick
+        if (record.startTick == curTick() &&
```

Comment:
Shouldn't this line be `record.startTick <= curTick()`?

Otherwise, you are only checking for *a single tick* not reserving the ALU for the ALU's latency.

Furthermore, the next line is comparing this to the number of things in the queue, which will *always* be `false`. There is no way `accesses_this_tick` could be more than the number of records in `accessQueue`.

Generally, this is difficult to understand code because it's not written in a way that is directly understandable from the purpose. It would be better to have an array of N entries (where N is the number of ALUs). When you reserve and entry, you put the time that the entry will be free in the location. When you are checking for an available entry, you go through and find if there are any entries with `free time < curTick()`. 

I could be misunderstanding this code, but I would encourage you to write the code to be more like the intended behavior. E.g., if you want to "reserve an ALU for `latency` time, then write code that does that instead of using a circular buffer and counts the number of units in use.

Reply:
@powerjg 
The intent of the code isn't to reserve ALUs themselves. As I'm assuming these ALUs are pipelined, multiple operations can be processed in a single ALU at once. _accessQueue doesn't keep track of ALUs_, but rather which cache lines are currently being processed..

Rather, it is to:
1. Restrict the number of ALU operations that can start in a cycle to _numALUs_
    - This is being accomplished in lines 71 to 74 (However, I was wrong in switching numALUs to accessQueue.size(), I'll switch that back as soon as I can)
2. Make sure that no ALU operation to a line already being processed can start until the previous operation has completed
    - This is accomplished in lines 77 to 79

This should model a group of numALUs pipelined ALUs that are accessLatency ticks deep, where a new operation can begin in each ALU every cycle.

Now that I think about it, however, I'd need to double-check that resource constraints for the same cycle are checked on the same tick. If they aren't I would need to change `curTick` to some sort of `curCycle`

Reply:
I see... I misunderstood what you were trying to model. My bad :).

It would be good to update the comments to make this more clear.

Reply:
@powerjg I've added additional comments and made the correction I mentioned.

Reply:
Thanks for the updates!